algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py

```diff
@@ -0,0 +1,177 @@
+                        num_batches: Optional[int] = None,
+                        repeat_final_dataset: bool = False):
+    per_device_batch_size = int(global_batch_size / N_GPUS)
+
+    # The input pipeline has to be created in all processes, because
+    # self._tokenizer has to be available in every process.
+    np_iter = super().build_input_queue(data_rng,
+                                        split,
+                                        data_dir,
+                                        global_batch_size,
+                                        num_batches,
+                                        repeat_final_dataset)
+    while True:
+      # Only iterate over tf input pipeline in one Python process to
+      # avoid creating too many threads.
+      if RANK == 0:
+        batch = next(np_iter)  # pylint: disable=stop-iteration-return
+        tensor_list = []
+        for key, value in batch.items():
+          tensor = torch.as_tensor(value, dtype=torch.float32, device=DEVICE)
+          tensor_list.append(tensor)
+          batch[key] = (
+              tensor[0] if USE_PYTORCH_DDP else tensor.view(
+                  -1, value.shape[-1]))
+        # Send batch to other devices when using DDP.
+        if USE_PYTORCH_DDP:
+          # During eval, the batch size of the remainder might be different.
+          if split != 'train':
+            per_device_batch_size = torch.tensor(
+                len(batch['inputs']), dtype=torch.int32, device=DEVICE)
+            dist.broadcast(per_device_batch_size, src=0)
+          dist.broadcast(torch.cat(tensor_list, dim=-1), src=0)
+      else:
+        # During eval, the batch size of the remainder might be different.
+        if split != 'train':
+          per_device_batch_size = torch.empty((1,),
+                                              dtype=torch.int32,
+                                              device=DEVICE)
+          dist.broadcast(per_device_batch_size, src=0)
+        tensor = torch.empty((3, N_GPUS, per_device_batch_size, 39),
+                             dtype=torch.float32,
+                             device=DEVICE)
+        dist.broadcast(tensor, src=0)
+        # Note that the order of the keys is important.
+        # tensors = tensor.split([[39, 1]], dim=-1)
+        keys = ['inputs', 'weights', 'targets']
+        batch = {}
+        for key, n in zip(keys, range(3)):
+          batch[key] = tensor[n][RANK]
+      yield batch
+
+  def _eval_batch(self, params, batch):
+    logits, _ = self.model_fn(
+        params,
+        batch,
+        model_state=None,
+        mode=spec.ForwardPassMode.EVAL,
+        rng=None,
+        update_batch_norm=False)
+    per_example_losses = metrics.per_example_sigmoid_binary_cross_entropy(
+        logits, batch['targets'])
+    batch_loss_numerator = torch.sum(per_example_losses)
+    batch_loss_denominator = np.sum(batch['weights'])
+    return np.asarray(batch_loss_numerator), batch_loss_denominator
```

Comment:
I'm probably missing something, but I think `np.asarray()` cannot be called on a GPU tensor?